### PR TITLE
Unlocks beneath

### PR DESCRIPTION
**Pull request by chcholman**
### Status
**READY**

### Description
This PR makes it possible to create unlock conditions for the Assignment Grade Earned condition, where student unlock new opportunities based on having earned less than a value or between a range of values on a particular assignment. 

### Todos
- [x] Add Tests
- [x] Update/Add Documentation
- [ ] Update Sample Data
- [x] Consider whether to add student-facing clarification of unlock above vs. below (and point value threshold)
- [x] Fix minimum and maximum threshold bug.

### Migrations
YES

### Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.
1. As an instructor, create a new Unlock Condition specifying an assignment, and selecting the condition "Grade Earned". 
2. Specify a maximum value. 
3. Go to the qualifying assignment and assess a student to have higher than that value - confirm that they have _not_ unlocked the associated assignment. 

### Impacted Areas in Application
List general components of the application that this PR will affect:

* Unlock conditions
* Assignment show page

======================

### CODE REVIEW

**Comment by maskib** [comment]
@ebarroso7 @chcholman The description in this PR makes it sound like we also want to allow unlocking beneath a student's overall score in the course, but that doesn't look implemented. Is this something we want to add?

**Comment by maskib** [comment]
@chcholman @ebarroso7 I also think that as part of this work it would be worth clarifying on the assignments tab on the student side of the app whether or not an assignment is unlocked by earning more, or earning less than a given point threshold, and also display what that threshold is. Currently, it just says "earn a grade", as in the screenshot below:

Screen Shot 2020-04-30 at 12 04 20 PM

**Comment by chcholman** [comment]
> @ebarroso7 @chcholman The description in this PR makes it sound like we also want to allow unlocking beneath a student's overall score in the course, but that doesn't look implemented. Is this something we want to add?

Clarified to note that this only applies to Assignment Grades earned, not overall course grade - that is an awesome feature but much larger.

**Comment by maskib** [comment]
@chcholman @ebarroso7 It is possible to specify both a maximum and minimum score in the unlock condition, but if you do so, you get a 500 error when attempting to grade the key assignment for this condition. There is an associated rollbar error: 
https://rollbar.com/gradecraft/gradecraft-development/items/5280/?person_page=0&#traceback

File: app/models/unlock_condition.rb [hunk 1]
```diff
@@ -227,15 +233,19 @@ def check_learning_objective_condition(student)
   def check_grade_earned_condition(student)
```
Comment (ebarroso7) [inline app/models/unlock_condition.rb:233]: @chcholman I think this is the method that is causing the issue Michael triggered. The rollbar error line number corresponds to the master-efs branch so if we look at that line number (273) in this file on this branch, its this method: `check_if_grade_earned_is_between_condition_values` 

When I try to trace this method I'm not seeing any if conditional statements that would catch a condition value range and condition date like: `if min_condition_value? && max_condition_value? && condition_date?`.  Do you think something like that is necessary ?  Happy to take a stab at it if you're busy (:

**Comment by maskib** [comment]
@ebarroso7 Looks pretty good! One last question about it - if you set the maximum points to be a lower value than the minimum points, then nobody will be able to achieve the unlock. Is this alright with us to allow?

**Comment by maskib** [comment]
@ebarroso7 Looking good! One questions to clarify - there are three "State Achieved" settings for assignment based unlock. This language update only seems to apply to the "grade earned" state. You can still specify points thresholds with the other two states though, so I want to figure out if that's just a UX inconsistency or of the language update needs to be extended to the other states?

**Comment by chcholman** [comment]
> @ebarroso7 @chcholman Looking good! One questions to clarify - there are three "State Achieved" settings for assignment based unlock. This language update only seems to apply to the "grade earned" state. You can still specify points thresholds with the other two states though, so I want to figure out if that's just a UX inconsistency or of the language update needs to be extended to the other states?

Great catch @maskib - this was inconsistency, and I've added the conditional such that those fields should only show if you've selected Grade Earned.

File: app/models/unlock_condition.rb [hunk 2]
```diff
@@ -100,7 +100,7 @@ def condition_state_do
       "Submit"
     elsif condition_state == "Grade Earned" && min_condition_value.present? && max_condition_value.present?
       "Score between #{min_condition_value} and #{max_condition_value} points on"
-    elsif condition_state == "Grade Earned" && min_condition_value.present?
+    elsif condition_state == "Grade Earned" && min_condition_value > 0
```
[the remark this answers is not in the document]
Comment (chcholman) [reply app/models/unlock_condition.rb]: I've always hated that solution, what do you think about the alt fix I just posted?

**Pull request merged**